### PR TITLE
[SYCL-MLIR] Replace remaining references to `pointer2memref(llvm.addrspacecast(memref2pointer))`

### PR DESCRIPTION
**Pull request by whitneywhtsang**

### CODE REVIEW

File: polygeist/test/polygeist-opt/licm.mlir [hunk 1]
```diff
@@ -300,46 +300,45 @@ func.func @affine_for_hoist5(%arg0: memref<?x!sycl_accessor_1_f32_rw_gb, 4>) {
...
   %alloca = memref.alloca() : memref<1x!sycl_id_1>  
   %alloca_0 = memref.alloca() : memref<1x!sycl_id_1>
   %c64_i64 = arith.constant 64 : i64
   %cst = arith.constant 1.000000e+01 : f32
 
-  %0 = "polygeist.memref2pointer"(%alloca) : (memref<1x!sycl_id_1>) -> !llvm.ptr<!sycl_id_1>
-  %1 = llvm.addrspacecast %0 : !llvm.ptr<!sycl_id_1> to !llvm.ptr<!sycl_id_1, 4>
-  %2 = "polygeist.pointer2memref"(%1) : (!llvm.ptr<!sycl_id_1, 4>) -> memref<?x!sycl_id_1, 4>
-  sycl.constructor @id(%2, %c64_i64) {MangledFunctionName = @_ZN4sycl3_V12idILi1EEC1ILi1EEENSt9enable_ifIXeqT_Li1EEmE4typeE} : (memref<?x!sycl_id_1, 4>, i64)
+  %0 = memref.cast %alloca : memref<1x!sycl_id_1> to memref<?x!sycl_id_1>
```
Comment (sommerlukas) [inline polygeist/test/polygeist-opt/licm.mlir:331]: Why do we need to cast the static shape information away here?

Comment (whitneywhtsang) [reply polygeist/test/polygeist-opt/licm.mlir:331]: We usually generate this cast, but got canonicalized away with `polygeist.memref2pointer`.

Comment (sommerlukas) [reply polygeist/test/polygeist-opt/licm.mlir:331]: Does the `sycl.constructor` not accept `memref<1x!sycl_id_1>`, i.e. memrefs with a static size? If so, why is that?

Comment (whitneywhtsang) [reply polygeist/test/polygeist-opt/licm.mlir:331]: It is a strict translation to the associated function call, which that function takes `memref<?x!sycl_id_1, 4>`. Are you suggesting to abstract the address space and memref shape from `sycl.constructor`?

**Review by etiotto — APPROVED** [review]

File: polygeist/test/polygeist-opt/licm.mlir (hunk 1 above)
Comment (sommerlukas) [reply polygeist/test/polygeist-opt/licm.mlir:331]: I'm mainly interested in the memref shape here. 

From my understanding of the `memref` type (and the `sycl.constructor` operation definition), I do not see a reason why the constructor or the associated function call would not accept a memref with a statically known shape, if they accept a memref with dynamic shape. After all, the static shape also guarantees that there is at least enough space for one element (the element to construct) in the memory buffer referenced by the memref. 

Is this related to the later lowerings expecting the parameters associated with dynamically shaped memrefs?

Comment (whitneywhtsang) [reply polygeist/test/polygeist-opt/licm.mlir:331]: Yup, `@_ZN4sycl3_V12idILi1EEC1ILi1EEENSt9enable_ifIXeqT_Li1EEmE4typeE` expects dynamically shaped memref.

Comment (sommerlukas) [reply polygeist/test/polygeist-opt/licm.mlir:331]: Ok, I see, thanks for the explanation!

**Review by sommerlukas — APPROVED** [review]

**Pull request merged**